heareval/tasks/speech_commands.py

```diff
@@ -2,38 +2,183 @@
 """
-# from heareval.tasks.coughvid import DownloadCorpus, ExtractCorpus
 import os
-import subprocess
+import re
+from functools import partial
+from glob import glob
 
+import heareval.tasks.config.speech_commands as config
 import luigi
 import numpy as np
-
-import heareval.tasks.config.speech_commands as config
-import heareval.tasks.util.audio as audio_util
-import heareval.tasks.util.luigi as luigi_util
-from heareval.tasks.util.luigi import DownloadCorpus, WorkTask, ExtractArchive
-
+import pandas as pd
+from heareval.tasks.util.luigi import (
+    PROCESSMETADATACOLS,
+    DownloadCorpus,
+    ExtractArchive,
+    FinalizeCorpus,
+    MetadataVocabulary,
+    MonoWavTrimCorpus,
+    ResampleSubCorpus,
+    SplitTrainTestCorpus,
+    SplitTrainTestMetadata,
+    SubsampleCorpus,
+    WorkTask,
+    ensure_dir,
+    filename_to_int_hash,
+    slugify_file_name,
+    which_set,
+)
 
 # Set the task name for all WorkTasks
 WorkTask.task_name = config.TASKNAME
 
 
+class ExtractArchive(ExtractArchive):
+    def requires(self):
+        return {
+            "download": DownloadCorpus(url=config.DOWNLOAD_URL, outfile="corpus.tar.gz")
+        }
+
+
+class ConfigureProcessMetaData(WorkTask):
+    """
+    This config is data dependent and has to be set for each data
+    """
+
+    def requires(self):
+        return ExtractArchive(infile="corpus.tar.gz")
+
+    def run(self):
+        # Get relative path of the audio files
+        # This file can also be built with the metadata file for the dataset
+        # in case the metadata is provided
+        process_metadata = pd.DataFrame(
+            glob(os.path.join(self.requires().workdir, "[!_]*/*.wav")),
+            columns=["relpath"],
+        )
+
+        # Get unique slug for each file
+        # In this case the base name is not excluded since same person have made dataset
+        # for different category
+        # This is something which is data dependent and should be handled here
+        process_metadata["slug"] = (
+            process_metadata["relpath"]
+            # Dont exclude the base name. Apply sluggify on all
+            .apply(partial(os.path.relpath, start=self.requires().workdir)).apply(
+                slugify_file_name
+            )
+        )
+
+        # Hash the field id rather than the full path.
+        # This hashing is specific to the dataset and should be done here
+        # In this case we take the slug and remove the -nohash- as described
+        # This nohash removal allows for the speech of similar person to be in the
+        # same dataset. Such type of data specific requirements might be there.
+        # in the readme of google speech commands. we want to keep similar people
+        # in the same group - test or train or val
+        process_metadata["filename_hash"] = (
+            process_metadata["slug"]
+            .apply(lambda relpath: re.sub(r"-nohash-.*$", "", relpath))
+            .apply(filename_to_int_hash)
+        )
```

Comment:
Google speech command also works by this method to sample the train test groups. They remove the no-hash part and then hash the file so that same speaker person will have a same hash and so that they fall in the same partition(`train` `test` or `val`). Its just that here i am applying the hash on the slug which is `label+filename minus the hash part`(reason being that similar persons have participated in the data generation for each label, but this is a choice). Can change it to the way done in the Google speech command which is something like `filename minus the hash part`. Is this strategy of doing everything in the `Configure process metadata` correct? The main goal here being that we can configure these `process level metadata` which are dependent on the `dataset` we are making the `pipeline` for and then we can plug in generic tasks which will probably take other task folders and this `process metadata` as input to produce subsequent output. @jorshi We can change the hash part for doing the split similar to the google speech command to get similar test train split and for the stratified subsampling, we can subsample this config file here so that the `SubSampleCorpus` can use this subsampled `process metadata`. This is how we can think about it in this setup. Please let me know if this is a good way?